src/utils/factory.hpp

```diff
@@ -0,0 +1,91 @@
+/*
+ * Authors: M. Kristien
+ * Organisation: HYPED
+ * Date: Dec 2019
+ * Description:
+ *
+ *    Copyright 2019 HYPED
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#ifndef UTILS_FACTORY_HPP_
+#define UTILS_FACTORY_HPP_
+/**
+ * @brief An abstract templated factory to provide mapping from string names
+ * of abstract interfaces to functions creating the implementations of the interfaces.
+ *
+ * The factory is templated with the interface class. This means the factory is created
+ * individually for different interfaces. Different implementations of the same interface
+ * use the same factory.
+ */
+
+#include <map>
+#include <string>
+
+#include "utils/interfaces.hpp"
+#include "utils/logger.hpp"
+
+using std::string;
+
+namespace hyped {
+namespace utils {
+
+template<class Interface>
+class Factory {
+  typedef Interface* (*InterfaceCreator)();
+  typedef std::map<string, InterfaceCreator> Mapping;
```

Comment:
Is this going to be used in many places? It is only used in one place atm so seems a bit pointless and actually hurts the readability rather than improving it because when I see the one usage of `Mapping` I just need to go and search for what it actually is.

Reply:
I just do not like writing:
```
  // Factory member variables
  Logger  log_;
  std::map<string, InterfaceCreator> mapping_;
```
Maybe the typedef could be closer to the member variable? (but that would violate class declaration order style rule)

Reply:
But you still have to write it :smile: You just add additional `Mapping; <...> Mapping` in between the type and the variable name. I would understand if it was used many times but it seems pointless if it's just once. Moving the typedef down would make it easier to navigate but it would still seem equally pointless. Also, I think the styleguide allows moving it closer if you make the typedef private?

Considering the statement `Mapping mapping_;`, the `Mapping` type does not add any information on top of what's already in the variable name. In `std::map<string, InterfaceCreator> mapping_;`, on the other hand, the type actually adds useful information on top of what is implied by the variable name.

Reply:
Yes, I still need to write the long map type but the member variables can be aligned. Frankly, I do not care strongly about this. Which version do you prefer then (private/no typedef)?

One could also argue this is private and should be hidden from any user of this code. If only we could write these in some `.cpp` file so that anyone who includes this header does not see it at all.

Reply:
I prefer no typedef. There are only 2 or 3 instance variables anyway so I don't think alignment is very important. I think that forcing reader to jump to another location to understand what `mapping_` is hurts readability more. Also, it is not only about using the code but also about maintaining it, in which case one does look at the private parts as well.